pager/src/main/java/dev/chrisbanes/accompanist/pager/PagerState.kt

```diff
@@ -0,0 +1,396 @@
+    pageCount: Int,
+    currentPage: Int = 0,
+    @FloatRange(from = 0.0, to = 1.0) currentPageOffset: Float = 0f,
+) {
+    private var _pageCount by mutableStateOf(pageCount)
+    private var _currentPage by mutableStateOf(currentPage)
+    private val _currentPageOffset = mutableStateOf(currentPageOffset)
+    internal var pageSize by mutableStateOf(0)
+
+    /**
+     * The number of pages to display.
+     */
+    @get:IntRange(from = 0)
+    var pageCount: Int
+        get() = _pageCount
+        set(@IntRange(from = 0) value) {
+            _pageCount = value.coerceAtLeast(0)
+            currentPage = currentPage.coerceIn(0, lastPageIndex)
+        }
+
+    internal val lastPageIndex: Int
+        get() = (pageCount - 1).coerceAtLeast(0)
+
+    /**
+     * The index of the currently selected page.
+     *
+     * To update the scroll position, use [scrollToPage] or [animateScrollToPage].
+     */
+    @get:IntRange(from = 0)
+    var currentPage: Int
+        get() = _currentPage
+        private set(value) {
+            _currentPage = value.coerceIn(0, _pageCount)
+        }
+
+    /**
+     * The current offset from the start of [currentPage], as a fraction of the page width.
+     *
+     * To update the scroll position, use [scrollToPage] or [animateScrollToPage].
+     */
+    @get:FloatRange(from = 0.0, to = 1.0)
+    var currentPageOffset: Float
+        get() = _currentPageOffset.value
+        private set(value) {
+            _currentPageOffset.value = value.coerceIn(
+                minimumValue = 0f,
+                maximumValue = if (currentPage == lastPageIndex) 0f else 1f,
+            )
+        }
+
+    /**
+     * Represents the current selection state of a [Pager].
+     * Usually read from [PagerState.selectionState].
+     */
+    enum class SelectionState {
+        /**
+         * Indicates that the pager is in an idle, settled state. The current page
+         * is fully in view and no animation is in progress.
+         */
+        Idle,
+
+        /**
+         * Indicates that the pager is currently being dragged by the user.
+         */
+        Dragging,
+
+        /**
+         * Indicates that the pager is in the process of settling to a final position.
+         */
+        Settling
+    }
+
+    /**
+     * The current selection state.
+     */
+    var selectionState by mutableStateOf(SelectionState.Idle)
+        internal set
+
+    /**
+     * Animate (smooth scroll) to the given page.
+     *
+     * Cancels the currently running scroll, if any, and suspends until the cancellation is
+     * complete.
+     *
+     * @param page the page to snap to. Must be between 0 and [pageCount] (inclusive).
+     * @param pageOffset the percentage of the page width to offset, from the start of [page]
+     * @param initialVelocity Initial velocity in pixels per second, or `0f` to not use a start velocity.
+     */
+    suspend fun animateScrollToPage(
+        @IntRange(from = 0) page: Int,
```

Comment:
do we need to crash if user passed incorrect values as params?

Reply:
Good catch, added a load of `require` checking